specs/configuration_files_specification.md

```diff
@@ -0,0 +1,64 @@
+---
+title: Configuration Files Specification
+layout: posts
+version: 1
+---
+
+# Configuration Files Specification
+
+## Introduction
+
+Various specifications attempt to define configuration files and file formats. This
+specification establishes where these files should be looked for, in which
+order, and how precedence, masking, extensions and overrides work.
+
+The purpose of the rules defined here is to allow OS vendors to implement the
+hermetic-usr pattern, where all vendor files are shipped in the vendor tree itself
+(`/usr/`), including configuration files with system defaults, while allowing local
+or vendor overrides without modifying the original files, for easier management. This is
+especially beneficial for image-based deployments, where the vendor tree is read-only.
+
+These rules are derived from existing real-world usage from the [systemd
+project](https://github.com/systemd/systemd) and the [libeconf
+project](https://github.com/openSUSE/libeconf).
```

Comment:
i don't think libeconf implements what systemd implements btw. there are similarities, but it seems to be sufficiently different. they seem to map `/etc/foobar.conf` to `/usr/<vendor>/foobar.conf`, while we map `/etc/<package>/foobar.conf` to `/usr/lib/<package>/foobar.conf`.

Frankly, I don't think what libeconf does makes much sense? Or maybe their docs are just bad? dunno.

Reply:
Also, they don't do /run/?

Reply:
There is no 100% overlap, but that's fine, because interoperability is not needed. The point here is to have something to point developers to that supports the common patterns of /etc -> /usr, drop-ins and masking.

Reply:
The reason for mentioning libeconf is that it's a library that can be integrated anywhere, that people can just use. systemd doesn't provide anything like that.

Reply:
> Also, they don't do /run/?

yes, it's mentioned as optional here

Reply:
> Also, they don't do /run/?

Yes, I noticed that as well (see https://github.com/uapi-group/specifications/pull/80#discussion_r1336284179).

Reply:
> i don't think libeconf implements what systemd implements btw. there are similarities, but it seems to be sufficiently different. they seem to map `/etc/foobar.conf` to `/usr/<vendor>/foobar.conf`, while we map `/etc/<package>/foobar.conf` to `/usr/lib/<package>/foobar.conf`.

There is not only `/usr/lib`, but packages also use `/usr/share`  and other directories. Since our goal was to be able to use libeconf for as many existing packages as possible, we decided to not use a fix prefix, but let the application tell libeconf where it has the config files below `/usr`. `<vendor>`  for `lib/<package>`, `share/<package>` or whatever else they are using may not be the best "naming", but the usage of that name has historic reasons.
If somebody has a better naming idea, I have no idea to let the documentation change.

Reply:
> > Also, they don't do /run/?
> 
> Yes, I noticed that as well (see [#80 (comment)](https://github.com/uapi-group/specifications/pull/80#discussion_r1336284179)).

Support for `/run` was not necessary in the past for our use cases. But it's on my wishlist, there are only no free resources currently.
If we have a spec for configuration files, I see no problem to adjust libeconf to follow that spec.

Reply:
> > Also, they don't do /run/?
> 
> Yes, I noticed that as well (see [#80 (comment)](https://github.com/uapi-group/specifications/pull/80#discussion_r1336284179)).

Wasn't necessary in the past for our use cases, but it's on my wishlist to get closer to what systemd is doing.

Reply:
> > i don't think libeconf implements what systemd implements btw. there are similarities, but it seems to be sufficiently different. they seem to map `/etc/foobar.conf` to `/usr/<vendor>/foobar.conf`, while we map `/etc/<package>/foobar.conf` to `/usr/lib/<package>/foobar.conf`.
> 
> There is not only `/usr/lib`, but packages also use `/usr/share` and other directories. Since our goal was to be able to use libeconf for as many existing packages as possible, we decided to not use a fix prefix, but let the application tell libeconf where it has the config files below `/usr`. `<vendor>` for `lib/<package>`, `share/<package>` or whatever else they are using may not be the best "naming", but the usage of that name has historic reasons. If somebody has a better naming idea, I have no idea to let the documentation change.

Ok I'll leave the exact location under `/usr/` up to the implementation, it is a minor detail after all

Reply:
> > > Also, they don't do /run/?
> > 
> > 
> > Yes, I noticed that as well (see [#80 (comment)](https://github.com/uapi-group/specifications/pull/80#discussion_r1336284179)).
> 
> Support for `/run` was not necessary in the past for our use cases. But it's on my wishlist, there are only no free resources currently. If we have a spec for configuration files, I see no problem to adjust libeconf to follow that spec.

I'll leave it optional in the spec - it is a nice feature to have for users, however it is not necessary to achieve the main goals of hermetic-usr and stateless systems